### PR TITLE
Xet is on the Hub blog post

### PR DESCRIPTION
**Pull request by jsulz**
Congratulations! You've made it this far! Once merged, the article will appear at https://huggingface.co/blog. Official articles 
require additional reviews. Alternatively, you can write a community article following the process [here](https://huggingface.co/blog-explorers).

## Preparing the Article

You're not quite done yet, though. Please make sure to follow this process (as documented [here](https://github.com/huggingface/blog/tree/main?tab=readme-ov-file#how-to-write-an-article-)):

- [x] Add an entry to [_blog.yml](https://github.com/huggingface/blog/blob/main/_blog.yml).
- [x] Add a thumbnail. There are no requirements here, but there is a [template](https://github.com/huggingface/blog?tab=readme-ov-file#how-to-get-a-nice-responsive-thumbnail) if it's helpful.
- [x] Check you use a short title and blog path.
- [x] Upload any additional assets (such as images) to the Documentation Images [repo](https://huggingface.co/datasets/huggingface/documentation-images/tree/main/blog). This is to reduce bloat in the GitHub base repo when cloning and pulling. Try to have small images to avoid a slow or expensive user experience.
- [x] Add metadata (such as authors) to your `md` file. You can also specify `guest` or `org` for the authors.
- [x] Ensure the publication date is correct.
- [x] Preview the content. A quick way is to paste the markdown content in https://huggingface.co/new-blog. **Do not click publish**, this is just a way to do an early check.

Here is an example of a complete PR: https://github.com/huggingface/blog/pull/2382

## Getting a Review

Please make sure to get a review from someone on your team or a co-author.
Once this is done and once all the steps above are completed, you should be able to merge.
There is no need for additional reviews if you and your co-authors are happy and meet all of the above.

Feel free to add @pcuenca as a reviewer if you want a final check. Keep in mind he'll be biased toward light reviews
(e.g., check for proper metadata) rather than content reviews unless explicitly asked.

### CODE REVIEW

**Comment by jsulz** [comment]
@rajatarya @ylow here's the migration analysis blog post.

I am most skeptical of the `Takeaways` and `Ready. Xet. Go.` sections but any and all feedback is welcome!

**_Aside:_** 
When I put the CTA into https://huggingface.co/new-blog and do a preview it looks like the screenshots below, but I'm not 100% certain it will show up like that. Will probably ask for a sanity check from another reviewer before merging anyway and ask about it then. 

Screenshot 2025-03-14 at 1 25 11 PM
Screenshot 2025-03-14 at 1 25 20 PM

File: xet-on-the-hub.md [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+<figure class="image text-center">
+    <img src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/xet-on-the-hub/pod-load-imbalance.png" alt="Pod load imbalance">
+    <figcaption>Balanced view of downloaded and sent data with reduced throughput across the board.</figcaption>
+</figure>
+
+To keep a pod from getting in this state, we put a limit to the number of concurrent uploads each machine accepts before rejecting requests that, when reached, pushes requests on to other pods in the cluster. If all pods in the cluster get in this state, then our autoscaling policies kick in. Longer-term solutions might include forcing disk syncs, removing the on-disk buffering, or switching to ephemeral storage with better throughput. 
+
+## Takeaways
+
+These two issues led to important architectural improvements, but they weren’t the only challenges the team tackled in the weeks following the migration:
```
Comment (ylow) [inline xet-on-the-hub.md:159]: I might omit the detailed bullet points and just go something like:
 - They weren’t the only challenges the team tackled in the weeks following the migration. We also addressed other performance issues, a memory leak (which we fixed but still can't root cause), and on advise from AWS changed from ALB to NLB to improve scalability.

File: xet-on-the-hub.md [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+
+* No test environment can simulate user behavior at scale. Even after careful integration work and Hugging Face employees testing the infrastructure for months, corner cases only surfaced once we funneled real usage through the system.  
+* By staging migrations incrementally and uncovering these issues before more traffic was on the infrastructure, we avoided downtime and disruption. Managing a fraction of traffic and storage made it relatively easy in comparison to if all the Hub was on Xet from day one.   
+* The infrastructure and system design was hardened iteratively over weeks. Every future byte and network request on Xet will receive the benefits of these lessons. 
+
+In short, real-world load was essential for exposing these challenges, and incremental migrations let us tackle them safely. That’s how we’re ensuring Xet-backed repositories scale reliably for the entire Hugging Face community with faster uploads, fewer bottlenecks, and minimal disruption all the way.
+
+### Ready. Xet. Go.
+
+With the initial migration complete, Xet is now on the Hugging Face Hub. We want to open up access more broadly, so you can feel the ease and speed of Xet storage.
```
Comment (ylow) [inline xet-on-the-hub.md:175]: "With the initial migration complete, Xet is now on the Hugging Face Hub!"

(Can drop "We want to open up access more broadly, so you can feel the ease and speed of Xet storage.", that gets repeated with the next paragraph)

File: xet-on-the-hub.md [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+<figure class="image text-center">
+    <img src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/xet-on-the-hub/block-format-download-problem.png" alt="Graphical representation of block format issue">
+    <figcaption>Red blocks are unnecessarily streamed content while green blocks show the data CAS is actually requesting and ultimately sends.</figcaption>
+</figure>
+
+To fix this, we updated the block format to store chunk-length metadata, enabling CAS to download only what each request actually needs. This format change required coordinated updates across:
+
+* CAS APIs  
+* Block metadata format  
+* All 65,000+ blocks in S3
```
Comment (ylow) [inline xet-on-the-hub.md:133]: "* 2^16 + 1 blocks on S3 (this is a coincidence I swear)"

File: _blog.yml [hunk 4]
```diff
@@ -5683,4 +5683,16 @@
...
+  title: "Xet is on the Hub"
+  author: jsulz
+  thumbnail: /blog/assets/xet-on-the-hub/thumbnail.png
+  date: March 17, 2025
+  tags:
+    - dedupe
+    - storage
+    - content defined chunking
+    - infrastructure
+    - content addressed store
```
Comment (julien-c) [inline _blog.yml:5698]:
```suggestion
    - xet
    - infra
```
maybe an explicit xet tag instead, but we would need to update the previous posts then

File: xet-on-the-hub.md [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+   * Provides the final persisted layer of data referenced by the CAS
+
+Before moving to production, the system was launched into an ephemeral environment, where we built a [steel thread of functionality](https://www.rubick.com/steel-threads/) to enable uploading and downloading via the `huggingface_hub`. The video below shows our progress one month in: 
+
+<video alt="Steel thread of Xet storage" controls>
+    <source src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/xet-on-the-hub/steel-thread.mp4" type="video/mp4">
+</video>
+
+After the heady highs of a quick proof-of-concept, the team settled into battling gnarly integration points (privacy, backward compatibility, fragmentation, etc) with the complex ecosystem that is the Hub. Eventually, the infrastructure moved into production for Hugging Face employees. With real usage now rolling in, we moved forward with the first large-scale migration.
```
Comment (julien-c) [inline xet-on-the-hub.md:85]:
```suggestion
After the heady highs of a quick proof-of-concept, the team settled into battling gnarly integration points (privacy, backward compatibility, fragmentation, etc) with the complex ecosystem that is the Hub. Eventually, the infrastructure moved into production for Hugging Face team members. With real usage now rolling in, we moved forward with the first large-scale migration.
```

File: xet-on-the-hub.md [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+
+* **Memory Leak** \- In the first few days after the migration, we noticed CAS memory usage increasing in a stepwise pattern, hinting at a leak which was fixed after investigation.
+* **Slow LFS Bridge Downloads** \- While investigating prolonged download times (a reminder to always keep an eye on tail latencies), we uncovered particularly troublesome cases. Interestingly, fixing the memory leak also resolved these slow downloads.  
+* **Switch from ALB to NLB** \- We eventually migrated the LFS Bridge from an [Application Load Balancer](https://docs.aws.amazon.com/elasticloadbalancing/latest/application/introduction.html) (ALB) to a [Network Load Balancer](https://docs.aws.amazon.com/elasticloadbalancing/latest/network/introduction.html) (NLB), improving scalability during bursts of traffic. The DNS cutover to the NLB was performed with zero downtime.
+
+Looking back, these issues underscore a set of critical lessons: 
+
+* No test environment can simulate user behavior at scale. Even after careful integration work and Hugging Face employees testing the infrastructure for months, corner cases only surfaced once we funneled real usage through the system.
```
Comment (julien-c) [inline xet-on-the-hub.md:167]:
```suggestion
* No test environment can simulate user behavior at scale. Even after careful integration work and Hugging Face team members testing the infrastructure for months, corner cases only surfaced once we funneled real usage through the system.  
```

File: xet-on-the-hub.md [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+
+In short, real-world load was essential for exposing these challenges, and incremental migrations let us tackle them safely. That’s how we’re ensuring Xet-backed repositories scale reliably for the entire Hugging Face community with faster uploads, fewer bottlenecks, and minimal disruption all the way.
+
+### Ready. Xet. Go.
+
+With the initial migration complete, Xet is now on the Hugging Face Hub. We want to open up access more broadly, so you can feel the ease and speed of Xet storage.
+
+<p align="center">
+    <img src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/xet-on-the-hub/migration.gif" alt="Xet speed" width=40%>
+</p>
```
Comment (julien-c) [inline xet-on-the-hub.md:179]: love this

**Review by julien-c — APPROVED** [review]
really nice.

This makes me think, maybe we should have a hf.co/xet/status or similar page where everyone could (publicly) see which % of all HF repos are Xet-backed (and see an evolution over time). Could be quite cool/motivating!

File: xet-on-the-hub.md (hunk 1 above)
Comment (jsulz) [reply xet-on-the-hub.md]: Updated!

File: xet-on-the-hub.md (hunk 2 above)
Comment (jsulz) [reply xet-on-the-hub.md]: Dropped.

File: xet-on-the-hub.md (hunk 3 above)
Comment (jsulz) [reply xet-on-the-hub.md]: Love it

File: _blog.yml (hunk 4 above)
Comment (jsulz) [reply _blog.yml]: I agree that this post has a few too many tags 😅 (don't really have to include CDC or dedupe - it's more about storage components and infrastructure). 

Not sure about xet tag 🤔 was thinking with an infra/infrastructure tag we could have the system component (e.g., "storage") dictate organization of blog content (with the `xet-team` org tag pointing people to team-specific content), but also don't have strong opinions.

For now, I just removed the unnecessary tags, but happy to make more updates!

Comment (julien-c) [reply _blog.yml]: yes, no strong opinion

**Review by ylow — APPROVED** [review]

File: xet-on-the-hub.md [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+</figure>
+
+To keep a pod from getting in this state, we put a limit to the number of concurrent uploads each machine accepts before rejecting requests that, when reached, pushes requests on to other pods in the cluster. If all pods in the cluster get in this state, then our autoscaling policies kick in. Longer-term solutions might include forcing disk syncs, removing the on-disk buffering, or switching to ephemeral storage with better throughput. 
+
+### Migration Takeaways
+
+These two issues led to important architectural improvements, but they weren’t the only challenges the team tackled in the weeks following the migration. We also addressed other performance issues, a memory leak (which we fixed but still can't root cause), and on advise from AWS to migrate the LFS Bridge from an [Application Load Balancer](https://docs.aws.amazon.com/elasticloadbalancing/latest/application/introduction.html) (ALB) to a [Network Load Balancer](https://docs.aws.amazon.com/elasticloadbalancing/latest/network/introduction.html) (NLB) to improve scalability during bursts of traffic.
+
+Looking back, these issues underscore a set of critical lessons:
```
Comment (rajatarya) [inline xet-on-the-hub.md:159]:
```suggestion
### Lessons learned
```

File: xet-on-the-hub.md [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+
+### Migration Takeaways
+
+These two issues led to important architectural improvements, but they weren’t the only challenges the team tackled in the weeks following the migration. We also addressed other performance issues, a memory leak (which we fixed but still can't root cause), and on advise from AWS to migrate the LFS Bridge from an [Application Load Balancer](https://docs.aws.amazon.com/elasticloadbalancing/latest/application/introduction.html) (ALB) to a [Network Load Balancer](https://docs.aws.amazon.com/elasticloadbalancing/latest/network/introduction.html) (NLB) to improve scalability during bursts of traffic.
+
+Looking back, these issues underscore a set of critical lessons: 
+
+* No test environment can simulate user behavior at scale. Even after careful integration work and Hugging Face team members testing the infrastructure for months, corner cases only surfaced once we funneled real usage through the system.
```
Comment (rajatarya) [inline xet-on-the-hub.md:161]:
```suggestion
* **There is nothing like PROD**: No test environment can simulate user behavior at scale. Even after careful integration work and Hugging Face team members testing the infrastructure for months, corner cases only surfaced once we funneled real usage through the system.  
```

File: xet-on-the-hub.md [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+
+These two issues led to important architectural improvements, but they weren’t the only challenges the team tackled in the weeks following the migration. We also addressed other performance issues, a memory leak (which we fixed but still can't root cause), and on advise from AWS to migrate the LFS Bridge from an [Application Load Balancer](https://docs.aws.amazon.com/elasticloadbalancing/latest/application/introduction.html) (ALB) to a [Network Load Balancer](https://docs.aws.amazon.com/elasticloadbalancing/latest/network/introduction.html) (NLB) to improve scalability during bursts of traffic.
+
+Looking back, these issues underscore a set of critical lessons: 
+
+* No test environment can simulate user behavior at scale. Even after careful integration work and Hugging Face team members testing the infrastructure for months, corner cases only surfaced once we funneled real usage through the system.  
+* By staging migrations incrementally and uncovering these issues before more traffic was on the infrastructure, we avoided downtime and disruption. Managing a fraction of traffic and storage made it relatively easy in comparison to if all the Hub was on Xet from day one.
```
Comment (rajatarya) [inline xet-on-the-hub.md:162]:
```suggestion
* **Migrations simulate real traffic**: By staging migrations incrementally and uncovering these issues before more traffic was on the infrastructure, we avoided downtime and disruption. Managing a fraction of traffic and storage made it relatively easy in comparison to if all the Hub was on Xet from day one.   
```

File: xet-on-the-hub.md [hunk 11]
```diff
@@ -0,0 +1,188 @@
...
+
+Looking back, these issues underscore a set of critical lessons: 
+
+* No test environment can simulate user behavior at scale. Even after careful integration work and Hugging Face team members testing the infrastructure for months, corner cases only surfaced once we funneled real usage through the system.  
+* By staging migrations incrementally and uncovering these issues before more traffic was on the infrastructure, we avoided downtime and disruption. Managing a fraction of traffic and storage made it relatively easy in comparison to if all the Hub was on Xet from day one.   
+* The infrastructure and system design was hardened iteratively over weeks. Every future byte and network request on Xet will receive the benefits of these lessons.
```
Comment (rajatarya) [inline xet-on-the-hub.md:163]:
```suggestion
* **Compound the learning**: The infrastructure and system design was hardened iteratively over weeks. Every future byte and network request on Xet will receive the benefits of these lessons. 
```

**Review by rajatarya — APPROVED** [review]
Minor comments - I think making the Lessons Learned easier to grok is worth it.

Overall this looks GREAT - excited for it to go out!

**Comment by jsulz** [comment]
Excellent - thanks all. Will wake up early tomorrow and publish this and then do a round of comms :) 

@pcuenca if you have time, always happy to get a review from you as well, but no worries if not.

File: xet-on-the-hub.md [hunk 12]
```diff
@@ -0,0 +1,188 @@
...
+</div>
+
+Over the past few weeks, Hugging Face’s [Xet Team](https://huggingface.co/xet-team) took a major step forward by [migrating the first Model and Dataset repositories off LFS and to Xet storage](https://huggingface.co/posts/jsulz/911431940353906).
+
+This marks one of many steps to [fulfill Hugging Face’s vision for the Hub](https://huggingface.co/blog/xethub-joins-hf) by empowering AI builders to build, iterate, and collaborate more effectively on massive models and datasets. If you're interested in deeper dives on the technology itself, check out the following posts:
+* [From Files to Chunks: Improving Hugging Face Storage Efficiency](https://huggingface.co/blog/from-files-to-chunks)
+* [Rearchitecting Hugging Face Uploads and Downloads](https://huggingface.co/blog/rearchitecting-uploads-and-downloads)
+* [From Chunks to Blocks: Accelerating Uploads and Downloads on the Hub](https://huggingface.co/blog/from-chunks-to-blocks)
+
+But this post isn't about the core technology. It's a a behind-the-scenes view of getting Xet on the Hub; taking you through our proof-of-concept to the first migration of repositories.
```
Comment (Vaibhavs10) [inline xet-on-the-hub.md:43]:
```suggestion
But this post isn't about the core technology. It's a behind-the-scenes view of getting Xet on the Hub; taking you through our proof-of-concept to the first migration of repositories. 
```

**Review by Vaibhavs10 — APPROVED** [review]
Nice! very informative post! 🤗

File: xet-on-the-hub.md [hunk 13]
```diff
@@ -0,0 +1,188 @@
...
+With the initial migration complete, Xet is now on the Hugging Face Hub.
+
+<p align="center">
+    <img src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/blog/xet-on-the-hub/migration.gif" alt="Xet speed" width=40%>
+</p>
+
+We’re finalizing an official integration with the `huggingface_hub` that will mean you get the benefits of Xet without any significant changes to your current workflow. To get going:
+
+1. **Join the Waitlist**  
+   * [Sign up here](https://huggingface.co/join-xet) with your user or organization account (if you have admin/write permissions).
```
Comment (tomaarsen) [inline xet-on-the-hub.md:178]: Should this be https://huggingface.co/join/xet instead of https://huggingface.co/join-xet?

@jsulz @rajatarya @ylow @julien-c

Comment (tomaarsen) [reply xet-on-the-hub.md:178]: If yes, I opened a PR with this proposal: #2747

Comment (jsulz) [reply xet-on-the-hub.md:178]: It should and thank you for the catch! PR approved and merged ❤️
